### PR TITLE
Add Text.rich constructor to Text widget

### LINKED ISSUE #13791: Text widget should have a constructor where you give a TextSpan
This would let people create const text widgets that are affected by the DefaultTextStyle.

### PR DESCRIPTION
**Pull request by jonahwilliams**
This allows a text widget to be constructed directly from a TextSpan.  The Tree of spans can each have their own styles while still inheriting from the default text styles (though these are not applied directly - I assume there is some magic here I am missing).  

Adds named constructor fromSpan to TextWidget
Adds member TextSpan span to TextWidget
Adds simple smoke test (any insight into testing these changes would be appreciated).

Fixes #13791

### CODE REVIEW

File: packages/flutter/lib/src/widgets/text.dart [hunk 1]
```diff
@@ -164,6 +164,11 @@ class DefaultTextStyle extends InheritedWidget {
 /// for example, to make the text bold while using the default font family and
 /// size.
 ///
+/// The [Text] widget can also be created directly from a [TextSpan]. Default text
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: not really "from", maybe "with"?

File: packages/flutter/lib/src/widgets/text.dart [hunk 2]
```diff
@@ -164,6 +164,11 @@ class DefaultTextStyle extends InheritedWidget {
...
+/// styles will not be provided to the span in this case, but they are used as default
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: i don't understand what this line means

File: packages/flutter/lib/src/widgets/text.dart [hunk 3]
```diff
@@ -210,16 +215,40 @@ class Text extends StatelessWidget {
     this.textScaleFactor,
     this.maxLines,
   }) : assert(data != null),
+       span = null,
        super(key: key);
 
+  /// Creates a text widget from a [TextSpan].
+  const Text.fromSpan(this.span, {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: "from" doesn't feel right here. It's not "from" a TextSpan any more than the other constructor is "from" a String. It's just that the child is a TextSpan.

Maybe `Text.rich`, following the RichText pattern?

File: packages/flutter/lib/src/widgets/text.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Let me try again

File: packages/flutter/lib/src/widgets/text.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Done

File: packages/flutter/lib/src/widgets/text.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: I combined this doc comment with the next line, since users no longer need to always use RichText.  RichText is still mentioned at the bottom, but I could add another reference to it earlier.

File: packages/flutter/lib/src/widgets/text.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Done

File: packages/flutter/lib/src/widgets/text.dart [hunk 4]
```diff
@@ -210,16 +215,40 @@ class Text extends StatelessWidget {
...
+
   /// The text to display.
+  ///
+  /// If constructed with the default constructor, `span` will be null.
   final String data;
 
+  /// The text to display as a span.
+  ///
+  /// If constructed from a [TextSpan], `data` will be null.
+  final TextSpan span;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: We should probably call this `textSpan` for clarity.

File: packages/flutter/lib/src/widgets/text.dart [hunk 5]
```diff
@@ -210,16 +215,40 @@ class Text extends StatelessWidget {
...
+    this.textScaleFactor,
+    this.maxLines,
+  }): assert(span != null),
+      data = null,
+      style = null,
+      super(key: key);
+
   /// The text to display.
+  ///
+  /// If constructed with the default constructor, `span` will be null.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: We usually frame this from the context of the constructors, so:

```
/// The text to display.
///
/// This will be null if a [textSpan] is provided instead.
```

File: packages/flutter/lib/src/widgets/text.dart [hunk 6]
```diff
@@ -210,16 +215,40 @@ class Text extends StatelessWidget {
...
   final String data;
 
+  /// The text to display as a span.
+  ///
+  /// If constructed from a [TextSpan], `data` will be null.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: (same comment here but in reverse)

File: packages/flutter/lib/src/widgets/text.dart [hunk 7]
```diff
@@ -284,7 +313,7 @@ class Text extends StatelessWidget {
       overflow: overflow ?? defaultTextStyle.overflow,
       textScaleFactor: textScaleFactor ?? MediaQuery.of(context, nullOk: true)?.textScaleFactor ?? 1.0,
       maxLines: maxLines ?? defaultTextStyle.maxLines,
-      text: new TextSpan(
+      text: span ?? new TextSpan(
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: if you do this, DefaultTextStyle isn't going to apply, which is going to defeat the entire purpose of this patch. :-)

File: packages/flutter/lib/src/widgets/text.dart [hunk 8]
```diff
@@ -294,7 +323,7 @@ class Text extends StatelessWidget {
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder description) {
     super.debugFillProperties(description);
-    description.add(new StringProperty('data', data, showName: false));
+    description.add(new StringProperty('data', data ?? span.toPlainText(), showName: false));
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: we should include span as its own property, that way it'll appear in the toStringDeep output nicely formatted.

File: packages/flutter/test/widgets/text_test.dart [hunk 9]
```diff
@@ -85,4 +85,22 @@ void main() {
...
+            const TextSpan(text: ' beautiful ', style: const TextStyle(fontStyle: FontStyle.italic)),
+            const TextSpan(text: 'world', style: const TextStyle(fontWeight: FontWeight.bold)),
+          ],
+        ),
+        textDirection: TextDirection.ltr,
+      ),
+    );
+
+    final RichText text = tester.firstWidget(find.byType(RichText));
+    expect(text, isNotNull);
```
Comment (Hixie) [inline packages/flutter/test/widgets/text_test.dart]: we should also check that DefaultTextStyle is applied

File: packages/flutter/lib/src/widgets/text.dart [hunk 10]
```diff
@@ -210,16 +215,40 @@ class Text extends StatelessWidget {
...
+  /// If constructed from a [TextSpan], `data` will be null.
+  final TextSpan span;
+
   /// If non-null, the style to use for this text.
   ///
   /// If the style's "inherit" property is true, the style will be merged with
   /// the closest enclosing [DefaultTextStyle]. Otherwise, the style will
   /// replace the closest enclosing [DefaultTextStyle].
+  ///
+  /// If constructed from a [TextSpan], the style will be null.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: we should probably allow style to be applied when you provide a TextSpan, the same way we do with text, just for completeness.

File: packages/flutter/lib/src/widgets/text.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Should I be specifically referring to the names of the constructor arguments, or just their types?

File: packages/flutter/lib/src/widgets/text.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Do we generally delegate to the existing debugFillProperty method (like style), or should I write my own PropertyDescription?

File: packages/flutter/lib/src/widgets/text.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Ohhhhh, okay, I think figured out how this is supposed to work

File: packages/flutter/test/widgets/text_test.dart (hunk 9 above)
Comment (jonahwilliams) [reply packages/flutter/test/widgets/text_test.dart]: Done

File: packages/flutter/lib/src/widgets/text.dart (hunk 8 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/text.dart]: TextSpan already implements Diagnosticable, so you should be able to do whatever RichText and RenderParagraph do already.

File: packages/flutter/lib/src/widgets/text.dart (hunk 5 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/text.dart]: As a general rule, we refer to the names of the properties, and name the constructor arguments the same as the properties.

File: packages/flutter/lib/src/widgets/text.dart [hunk 11]
```diff
@@ -164,8 +164,8 @@ class DefaultTextStyle extends InheritedWidget {
 /// for example, to make the text bold while using the default font family and
 /// size.
 ///
-/// To display text that uses multiple styles (e.g., a paragraph with some bold
-/// words), use [RichText].
+/// The [Text] widget can also be created with a [TextSpan] to display text that
+/// uses multiple styles (e.g., a paragraph with some bold words).
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/text.dart]: maybe explicitly say `Using the [new Text.rich] constructor, the [Text] widget can also be created with a [TextSpan] ...`

(`[new Foo]` is how you refer to constructors in dartdocs)

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/widgets/text.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Needs work

File: packages/flutter/lib/src/widgets/text.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Done

File: packages/flutter/lib/src/widgets/text.dart (hunk 10 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/text.dart]: Done

**Comment by jonahwilliams** [comment]
Checked that diagnostics and default styles seem to be working as expected.
